src/App.tsx

```diff
@@ -1,24 +1,248 @@
+        setRequest(false);
+      });
+  }, []);
+
+  const idTimerRef = useRef<number>(0);
+
+  useEffect(() => {
+    if (idTimerRef.current) {
+      window.clearTimeout(idTimerRef.current);
+    }
+
+    idTimerRef.current = window.setTimeout(() => {
+      setErrorMessage('');
+    }, 3000);
+  }, [setErrorMessage]);
+
+  const handleAddTodos = (newTodo: Omit<Todo, 'id'>) => {
+    setLoadingId([0]);
+    setIsLoading(true);
+    setRequest(true);
+    addTodos(newTodo)
+      .then((createdTodo) => {
+        setTodo((currentTodo) => [...currentTodo, createdTodo]);
+        setTitle('');
+        setIsLoading(false);
+      })
+      .catch(() => {
+        setErrorMessage('Unable to add a todo');
+        setTimeout(() => setErrorMessage(''), 3000);
+      })
+      .finally(() => {
+        setTempTodo(null);
+        setIsLoading(false);
+        setRequest(false);
+      });
+    const temp: Todo = Object.assign(newTodo, { id: 0 });
+
+    setTempTodo(temp);
+  };
+
+  const handleDelete = (todoId: number) => {
+    setLoadingId([todoId]);
+    setIsLoaderActive(true);
+    deleteTodos(todoId)
+      .then(() => {
+        setTodo(prevTodos => prevTodos.filter(todo => todo.id !== todoId));
+        setErrorMessage('');
+      })
+      .catch(() => {
+        setErrorMessage('Unable to delete todo');
+      })
+      .finally(() => {
+        setLoadingId([]);
+        setIsLoading(false);
+        setIsLoaderActive(false);
+      });
+  };
+
+  const handleClearCompleted = () => {
+    const deletePromises = todos
+      .filter(({ completed }) => completed)
+      .map(({ id }) => deleteTodos(id));
+
+    Promise.all(deletePromises)
+      .then(() => {
+        return getTodos(USER_ID);
+      })
+      .then((updatedTodos) => {
+        setTodo(updatedTodos);
+      })
+      .catch(() => {
+        completedTodosCount = false;
+        setErrorMessage('Unable to delete a todo');
+      });
+  };
+
+  const handleDeleteUpdate = (todo: Todo, newTodoTitle: string) => {
+    updateTodo({
+      id: todo.id,
+      title: newTodoTitle,
+      userId: todo.userId,
+      completed: todo.completed,
+    })
+      .then(updatedTodo => {
+        setTodo(prevState => prevState.map(currentTodo => (
+          currentTodo.id !== updatedTodo.id
+            ? currentTodo
+            : updatedTodo
+        )));
+      })
+      .finally(() => {
+        setLoadingId([]);
+        setIsLoaderActive(false);
+      });
+  };
+
+  const handleToggleTodo = (todo: Todo) => {
+    setLoadingId((prevTodoId) => [...prevTodoId, todo.id]);
+
+    return updateTodo({
+      ...todo,
+      completed: !todo.completed,
+    })
+      .then(updatedTodo => {
+        setTodo(prevState => prevState.map(currentTodo => (
+          currentTodo.id !== updatedTodo.id
+            ? currentTodo
+            : updatedTodo
+        )));
+      })
+      .catch(() => {
+        setErrorMessage('Unable to toggle a todo');
+        throw new Error();
```

Comment:
The error text must be passed to the `Error` constructor class, otherwise it does not make much sense